File: src/php/ext/grpc/hhvm/ext_grpc.cpp
/*
 *
 * Copyright 2015 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

 #ifdef HAVE_CONFIG_H
    #include "config.h"
#endif

#include "call.h"
#include "channel.h"
#include "server.h"
#include "timeval.h"
#include "channel_credentials.h"
#include "call_credentials.h"
#include "server_credentials.h"
#include "completion_queue.h"
#include "version.h"

#include "hphp/runtime/ext/extension.h"
#include "hphp/runtime/vm/native-data.h"

using namespace HPHP;

namespace HPHP {

class GrpcExtension : public Extension
{
public:
    GrpcExtension(void) : Extension("grpc", HHVM_GRPC_VERSION) {}
    virtual void moduleInit(void)
    {
        /* Register call error constants */
        grpc_init();

        grpc_hhvm_init_channel_credentials();

        HHVM_RC_INT(Grpc\\CALL_OK, GRPC_CALL_OK);
        HHVM_RC_INT(Grpc\\CALL_ERROR, GRPC_CALL_ERROR);
        HHVM_RC_INT(Grpc\\CALL_ERROR_NOT_ON_SERVER, GRPC_CALL_ERROR_NOT_ON_SERVER);
        HHVM_RC_INT(Grpc\\CALL_ERROR_NOT_ON_CLIENT, GRPC_CALL_ERROR_NOT_ON_CLIENT);
        HHVM_RC_INT(Grpc\\CALL_ERROR_ALREADY_INVOKED, GRPC_CALL_ERROR_ALREADY_INVOKED);
        HHVM_RC_INT(Grpc\\CALL_ERROR_NOT_INVOKED, GRPC_CALL_ERROR_NOT_INVOKED);
        HHVM_RC_INT(Grpc\\CALL_ERROR_ALREADY_FINISHED, GRPC_CALL_ERROR_ALREADY_FINISHED);
        HHVM_RC_INT(Grpc\\CALL_ERROR_TOO_MANY_OPERATIONS, GRPC_CALL_ERROR_TOO_MANY_OPERATIONS);
        HHVM_RC_INT(Grpc\\CALL_ERROR_INVALID_FLAGS, GRPC_CALL_ERROR_INVALID_FLAGS);

        HHVM_RC_INT(Grpc\\WRITE_BUFFER_HINT, GRPC_WRITE_BUFFER_HINT);
        HHVM_RC_INT(Grpc\\WRITE_NO_COMPRESS, GRPC_WRITE_NO_COMPRESS);

        HHVM_RC_INT(Grpc\\STATUS_OK, GRPC_STATUS_OK);
        HHVM_RC_INT(Grpc\\STATUS_CANCELLED, GRPC_STATUS_CANCELLED);
        HHVM_RC_INT(Grpc\\STATUS_UNKNOWN, GRPC_STATUS_UNKNOWN);
        HHVM_RC_INT(Grpc\\STATUS_INVALID_ARGUMENT, GRPC_STATUS_INVALID_ARGUMENT);
        HHVM_RC_INT(Grpc\\STATUS_DEADLINE_EXCEEDED, GRPC_STATUS_DEADLINE_EXCEEDED);
        HHVM_RC_INT(Grpc\\STATUS_NOT_FOUND, GRPC_STATUS_NOT_FOUND);
        HHVM_RC_INT(Grpc\\STATUS_ALREADY_EXISTS, GRPC_STATUS_ALREADY_EXISTS);
        HHVM_RC_INT(Grpc\\STATUS_PERMISSION_DENIED, GRPC_STATUS_PERMISSION_DENIED);
        HHVM_RC_INT(Grpc\\STATUS_UNAUTHENTICATED, GRPC_STATUS_UNAUTHENTICATED);
        HHVM_RC_INT(Grpc\\STATUS_RESOURCE_EXHAUSTED, GRPC_STATUS_RESOURCE_EXHAUSTED);
        HHVM_RC_INT(Grpc\\STATUS_FAILED_PRECONDITION, GRPC_STATUS_FAILED_PRECONDITION);
        HHVM_RC_INT(Grpc\\STATUS_ABORTED, GRPC_STATUS_ABORTED);
        HHVM_RC_INT(Grpc\\STATUS_OUT_OF_RANGE, GRPC_STATUS_OUT_OF_RANGE);
        HHVM_RC_INT(Grpc\\STATUS_UNIMPLEMENTED, GRPC_STATUS_UNIMPLEMENTED);
        HHVM_RC_INT(Grpc\\STATUS_INTERNAL, GRPC_STATUS_INTERNAL);
        HHVM_RC_INT(Grpc\\STATUS_UNAVAILABLE, GRPC_STATUS_UNAVAILABLE);
        HHVM_RC_INT(Grpc\\STATUS_DATA_LOSS, GRPC_STATUS_DATA_LOSS);

        HHVM_RC_INT(Grpc\\OP_SEND_INITIAL_METADATA, GRPC_OP_SEND_INITIAL_METADATA);
        HHVM_RC_INT(Grpc\\OP_SEND_MESSAGE, GRPC_OP_SEND_MESSAGE);
        HHVM_RC_INT(Grpc\\OP_SEND_CLOSE_FROM_CLIENT, GRPC_OP_SEND_CLOSE_FROM_CLIENT);
        HHVM_RC_INT(Grpc\\OP_SEND_STATUS_FROM_SERVER, GRPC_OP_SEND_STATUS_FROM_SERVER);
        HHVM_RC_INT(Grpc\\OP_RECV_INITIAL_METADATA, GRPC_OP_RECV_INITIAL_METADATA);
        HHVM_RC_INT(Grpc\\OP_RECV_MESSAGE, GRPC_OP_RECV_MESSAGE);
        HHVM_RC_INT(Grpc\\OP_RECV_STATUS_ON_CLIENT, GRPC_OP_RECV_STATUS_ON_CLIENT);
        HHVM_RC_INT(Grpc\\OP_RECV_CLOSE_ON_SERVER, GRPC_OP_RECV_CLOSE_ON_SERVER);

        HHVM_RC_INT(Grpc\\CHANNEL_IDLE, GRPC_CHANNEL_IDLE);
        HHVM_RC_INT(Grpc\\CHANNEL_CONNECTING, GRPC_CHANNEL_CONNECTING);
        HHVM_RC_INT(Grpc\\CHANNEL_READY, GRPC_CHANNEL_READY);
        HHVM_RC_INT(Grpc\\CHANNEL_TRANSIENT_FAILURE, GRPC_CHANNEL_TRANSIENT_FAILURE);
        HHVM_RC_INT(Grpc\\CHANNEL_FATAL_FAILURE, GRPC_CHANNEL_SHUTDOWN);

        HHVM_MALIAS(Grpc\\Call, __construct, Call, __construct);
        HHVM_MALIAS(Grpc\\Call, startBatch, Call, startBatch);
        HHVM_MALIAS(Grpc\\Call, getPeer, Call, getPeer);
        HHVM_MALIAS(Grpc\\Call, cancel, Call, cancel);
        HHVM_MALIAS(Grpc\\Call, setCredentials, Call, setCredentials);

        HHVM_STATIC_MALIAS(Grpc\\CallCredentials, createComposite, CallCredentials, createComposite);
        HHVM_STATIC_MALIAS(Grpc\\CallCredentials, createFromPlugin, CallCredentials, createFromPlugin);

        HHVM_MALIAS(Grpc\\Channel, __construct, Channel, __construct);
        HHVM_MALIAS(Grpc\\Channel, getTarget, Channel, getTarget);
        HHVM_MALIAS(Grpc\\Channel, getConnectivityState, Channel, getConnectivityState);
        HHVM_MALIAS(Grpc\\Channel, watchConnectivityState, Channel, watchConnectivityState);
        HHVM_MALIAS(Grpc\\Channel, close, Channel, close);

        HHVM_STATIC_MALIAS(Grpc\\ChannelCredentials, setDefaultRootsPem, ChannelCredentials, setDefaultRootsPem);
        HHVM_STATIC_MALIAS(Grpc\\ChannelCredentials, createDefault, ChannelCredentials, createDefault);
        HHVM_STATIC_MALIAS(Grpc\\ChannelCredentials, createSsl, ChannelCredentials, createSsl);
        HHVM_STATIC_MALIAS(Grpc\\ChannelCredentials, createComposite, ChannelCredentials, createComposite);
        HHVM_STATIC_MALIAS(Grpc\\ChannelCredentials, createInsecure, ChannelCredentials, createInsecure);

        HHVM_MALIAS(Grpc\\Server, __construct, Server, __construct);
        HHVM_MALIAS(Grpc\\Server, requestCall, Server, requestCall);
        HHVM_MALIAS(Grpc\\Server, addHttp2Port, Server, addHttp2Port);
        HHVM_MALIAS(Grpc\\Server, addSecureHttp2Port, Server, addSecureHttp2Port);
        HHVM_MALIAS(Grpc\\Server, start, Server, start);

        HHVM_STATIC_MALIAS(Grpc\\ServerCredentials, createSsl, ServerCredentials, createSsl);

        HHVM_MALIAS(Grpc\\Timeval, __construct, Timeval, __construct);
        HHVM_MALIAS(Grpc\\Timeval, add, Timeval, add);
        HHVM_MALIAS(Grpc\\Timeval, subtract, Timeval, subtract);
        HHVM_STATIC_MALIAS(Grpc\\Timeval, compare, Timeval, compare);
        HHVM_STATIC_MALIAS(Grpc\\Timeval, similar, Timeval, similar);
        HHVM_STATIC_MALIAS(Grpc\\Timeval, now, Timeval, now);
        HHVM_STATIC_MALIAS(Grpc\\Timeval, zero, Timeval, zero);
        HHVM_STATIC_MALIAS(Grpc\\Timeval, infFuture, Timeval, infFuture);
        HHVM_STATIC_MALIAS(Grpc\\Timeval, infPast, Timeval, infPast);
        HHVM_MALIAS(Grpc\\Timeval, sleepUntil, Timeval, sleepUntil);

        Native::registerNativeDataInfo<TimevalData>(TimevalData::className().get());

        Native::registerNativeDataInfo<ServerCredentialsData>(ServerCredentialsData::s_className.get());
        Native::registerNativeDataInfo<ServerData>(ServerData::className().get());

        Native::registerNativeDataInfo<ChannelCredentialsData>(ChannelCredentialsData::s_className.get());
        Native::registerNativeDataInfo<ChannelData>(ChannelData::className().get());

        Native::registerNativeDataInfo<CallCredentialsData>(CallCredentialsData::s_className.get());
        Native::registerNativeDataInfo<CallData>(CallData::className().get());

        loadSystemlib();
    }

    virtual void moduleShutdown(void)
    {
        /*
        {
            Lock l1(s_global_channels_cache_mutex);
            for (auto it = s_global_channels_cache.globalChannelMap.begin();
                      it != s_global_channels_cache.globalChannelMap.end(); ++it)
            {
                grpc_channel_destroy(*it);
            }

            s_global_channels_cache.globalChannelMap.clear();
        }
        */

        grpc_shutdown();
    }

} s_grpc_extension;

HHVM_GET_MODULE(grpc);

} // namespace HPHP

File: src/php/ext/grpc/hhvm/ext_grpc.php
<?hh
/*
 *
 * Copyright 2015 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

namespace Grpc {

<<__NativeData("Grpc\\Call")>>
class Call {

  /**
   * Constructs a new instance of the Call class.
   * @param Channel $channel_obj The channel to associate the call with.
   *                             Must not be closed.
   * @param string $method The method to call
   * @param Timeval $deadline_obj The deadline for completing the call
   * @param string $host_override The host is set by user (optional)
   */
  <<__Native>>
  public function __construct(Channel $channel, string $method, Timeval $deadlineTimeval, ?string $host_override = null): void;

  /**
   * Start a batch of RPC actions.
   * @param array $array Array of actions to take
   * @return object Object with results of all actions
   */
  <<__Native>>
  public function startBatch(array $actions): object;

  /**
   * Get the endpoint this call/stream is connected to
   * @return string The URI of the endpoint
   */
  <<__Native>>
  public function getPeer(): string;

  /**
   * Cancel the call. This will cause the call to end with STATUS_CANCELLED
   * if it has not already ended with another status.
   * @return void
   */
  <<__Native>>
  public function cancel(): void;

  /**
   * Set the CallCredentials for this call.
   * @param CallCredentials $creds_obj The CallCredentials object
   * @return int The error code
   */
  <<__Native>>
  public function setCredentials(CallCredentials $credentials): int;
}

<<__NativeData("Grpc\\CallCredentials")>>
class CallCredentials {

  /**
   * Create composite credentials from two existing credentials.
   * @param CallCredentials $cred1_obj The first credential
   * @param CallCredentials $cred2_obj The second credential
   * @return CallCredentials The new composite credentials object
   */
  <<__Native>>
  public static function createComposite(CallCredentials $cred1, CallCredentials $cred2): CallCredentials;

  /**
   * Create a call credentials object from the plugin API
   * @param callable $fci The callback
   * @return CallCredentials The new call credentials object
   */
  <<__Native>>
  public static function createFromPlugin(mixed $callback): CallCredentials;

}

<<__NativeData("Grpc\\Channel")>>
class Channel {

  /**
   * Construct an instance of the Channel class. If the $args array contains a
   * "credentials" key mapping to a ChannelCredentials object, a secure channel
   * will be created with those credentials.
   * @param string $target The hostname to associate with this channel
   * @param array $args_array The arguments to pass to the Channel
   */
  <<__Native>>
  public function __construct(string $target, array<string, mixed> $args): void;

  /**
   * Get the endpoint this call/stream is connected to
   * @return string The URI of the endpoint
   */
  <<__Native>>
  public function getTarget(): string;

  /**
   * Get the connectivity state of the channel
   * @param bool $try_to_connect Try to connect on the channel (optional)
   * @return long The grpc connectivity state
   */
  <<__Native>>
  public function getConnectivityState(bool $try_to_connect = false): int;

  /**
   * Watch the connectivity state of the channel until it changed
   * @param long $last_state The previous connectivity state of the channel
   * @param Timeval $deadline_obj The deadline this function should wait until
   * @return bool If the connectivity state changes from last_state
   *              before deadline
   */
  <<__Native>>
  public function watchConnectivityState(int $last_state, Timeval $deadline_timeval): bool;

  /**
   * Close the channel
   * @return void
   */
  <<__Native>>
  public function close(): void;

}

<<__NativeData("Grpc\\ChannelCredentials")>>
class ChannelCredentials {

  /**
   * Set default roots pem.
   * @param string $pem_roots PEM encoding of the server root certificates
   * @return void
   */
  <<__Native>>
  public static function setDefaultRootsPem(string $pem_roots): void;

  /**
   * Create a default channel credentials object.
   * @return ChannelCredentials The new default channel credentials object
   */
  <<__Native>>
  public static function createDefault(): ChannelCredentials;

  /**
   * Create SSL credentials.
   * @param string $pem_root_certs PEM encoding of the server root certificates
   * @param string $pem_key_cert_pair.private_key PEM encoding of the client's
   *                                              private key (optional)
   * @param string $pem_key_cert_pair.cert_chain PEM encoding of the client's
   *                                             certificate chain (optional)
   * @return ChannelCredentials The new SSL credentials object
   */
  <<__Native>>
  public static function createSsl(?string $pem_root_certs = null,
    ?string $pem_key_cert_pair__private_key = null,
    ?string $pem_key_cert_pair__cert_chain = null): ChannelCredentials;

  /**
   * Create composite credentials from two existing credentials.
   * @param ChannelCredentials $cred1_obj The first credential
   * @param CallCredentials $cred2_obj The second credential
   * @return ChannelCredentials The new composite credentials object
   */
  <<__Native>>
  public static function createComposite(ChannelCredentials $cred1_obj, CallCredentials $cred2_obj): ChannelCredentials;

  /**
   * Create insecure channel credentials
   * @return null
   */
  <<__Native>>
  public static function createInsecure(): mixed;

}

<<__NativeData("Grpc\\Server")>>
class Server {

  /**
   * Constructs a new instance of the Server class
   * @param array $args_array The arguments to pass to the server (optional)
   */
  <<__Native>>
  public function __construct(?array $args = null): void;

  /**
   * Request a call on a server. Creates a single GRPC_SERVER_RPC_NEW event.
   * @return object (result object)
   */
  <<__Native>>
  public function requestCall(): object;

  /**
   * Add a http2 over tcp listener.
   * @param string $addr The address to add
   * @return bool True on success, false on failure
   */
  <<__Native>>
  public function addHttp2Port(string $addr): bool;

  /**
   * Add a secure http2 over tcp listener.
   * @param string $addr The address to add
   * @param ServerCredentials The ServerCredentials object
   * @return int port on success, 0 on failure
   */
  <<__Native>>
  public function addSecureHttp2Port(string $addr, ServerCredentials $serverCredentials): int;

  /**
   * Start a server - tells all listeners to start listening
   * @return void
   */
  <<__Native>>
  public function start(): void;

}

<<__NativeData("Grpc\\ServerCredentials")>>
class ServerCredentials {

  /**
   * Create SSL credentials.
   * @param string pem_root_certs PEM encoding of the server root certificates
   * @param string pem_private_key PEM encoding of the client's private key
   * @param string pem_cert_chain PEM encoding of the client's certificate chain
   * @return Credentials The new SSL credentials object
   */
  <<__Native>>
  public static function createSsl(string $pem_root_certs, string $pem_private_key, string $pem_cert_chain): object;

}

<<__NativeData("Grpc\\Timeval")>>
class Timeval {

  /**
   * Constructs a new instance of the Timeval class
   * @param long $microseconds The number of microseconds in the interval
   */
  <<__Native>>
  public function __construct(int $microseconds): void;

  /**
   * Adds another Timeval to this one and returns the sum. Calculations saturate
   * at infinities.
   * @param Timeval $other_obj The other Timeval object to add
   * @return Timeval A new Timeval object containing the sum
   */
  <<__Native>>
  public function add(Timeval $otherTimeval): Timeval;

  /**
   * Subtracts another Timeval from this one and returns the difference.
   * Calculations saturate at infinities.
   * @param Timeval $other_obj The other Timeval object to subtract
   * @return Timeval A new Timeval object containing the diff 
   */
  <<__Native>>
  public function subtract(Timeval $otherTimeval): Timeval;

  /**
   * Return negative, 0, or positive according to whether a < b, a == b,
   * or a > b respectively.
   * @param Timeval $a_obj The first time to compare
   * @param Timeval $b_obj The second time to compare
   * @return long
   */
  <<__Native>>
  public static function compare(Timeval $timevalA, Timeval $timevalB): int;

  /**
   * Checks whether the two times are within $threshold of each other
   * @param Timeval $a_obj The first time to compare
   * @param Timeval $b_obj The second time to compare
   * @param Timeval $thresh_obj The threshold to check against
   * @return bool True if $a and $b are within $threshold, False otherwise
   */
  <<__Native>>
  public static function similar(Timeval $timevalA, Timeval $timevalB, Timeval $thresholdTimeval): bool;

  /**
   * Returns the current time as a timeval object
   * @return Timeval The current time
   */
  <<__Native>>
  public static function now(): Timeval;

  /**
   * Returns the zero time interval as a timeval object
   * @return Timeval Zero length time interval
   */
  <<__Native>>
  public static function zero(): Timeval;

  /**
   * Returns the infinite future time value as a timeval object
   * @return Timeval Infinite future time value
   */
  <<__Native>>
  public static function infFuture(): Timeval;

  /**
   * Returns the infinite past time value as a timeval object
   * @return Timeval Infinite past time value
   */
  <<__Native>>
  public static function infPast(): Timeval;

  /**
   * Sleep until this time, interpreted as an absolute timeout
   * @return void
   */
  <<__Native>>
  public function sleepUntil(): void;
}

}
